{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metagenomic-based Diagnostic for Sepsis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Statements\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import GridSearchCV, StratifiedKFold, cross_val_score\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "\n",
    "cwd = Path.cwd()\n",
    "datasets = cwd / 'datasets'\n",
    "results = cwd / 'results'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Preprocessing\n",
    "Since we are using stratified kfold, a validation split is not necesssary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathogen</th>\n",
       "      <th>y</th>\n",
       "      <th>Bradyrhizobium</th>\n",
       "      <th>Rhodopseudomonas</th>\n",
       "      <th>Bosea</th>\n",
       "      <th>Afipia</th>\n",
       "      <th>Oligotropha</th>\n",
       "      <th>Variibacter</th>\n",
       "      <th>Methylobacterium</th>\n",
       "      <th>Methylorubrum</th>\n",
       "      <th>...</th>\n",
       "      <th>Rubinisphaera</th>\n",
       "      <th>Dictyoglomus</th>\n",
       "      <th>Pakpunavirus</th>\n",
       "      <th>Marinilactibacillus</th>\n",
       "      <th>Paludibacter</th>\n",
       "      <th>Nonagvirus</th>\n",
       "      <th>Halovivax</th>\n",
       "      <th>Phifelvirus</th>\n",
       "      <th>Planktothrix</th>\n",
       "      <th>Denitrobacterium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>13825.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2883.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>20476.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8183.0</td>\n",
       "      <td>480.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>9677.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>436.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>15211.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7081.0</td>\n",
       "      <td>2041.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>56586.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13850.0</td>\n",
       "      <td>1856.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>9392.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4187.0</td>\n",
       "      <td>199.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>septic</td>\n",
       "      <td>3466.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>481.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Streptococcus oralis</td>\n",
       "      <td>septic</td>\n",
       "      <td>17287.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>408.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>4006.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4456.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>3354.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows × 1058 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pathogen        y  Bradyrhizobium  Rhodopseudomonas  \\\n",
       "0                       none  healthy         13825.0             108.0   \n",
       "1                       none  healthy         20476.0             234.0   \n",
       "2                       none  healthy          9677.0              72.0   \n",
       "3                       none  healthy         15211.0             158.0   \n",
       "4                       none  healthy         56586.0             294.0   \n",
       "..                       ...      ...             ...               ...   \n",
       "279         Escherichia coli   septic          9392.0              54.0   \n",
       "280  Cryptococcus neoformans   septic          3466.0              40.0   \n",
       "281     Streptococcus oralis   septic         17287.0             142.0   \n",
       "282         Escherichia coli   septic          4006.0              39.0   \n",
       "283         Escherichia coli   septic          3354.0              49.0   \n",
       "\n",
       "     Bosea  Afipia  Oligotropha  Variibacter  Methylobacterium  Methylorubrum  \\\n",
       "0    130.0     0.0         12.0          0.0            2883.0           82.0   \n",
       "1     60.0    53.0         30.0          0.0            8183.0          480.0   \n",
       "2     17.0     8.0         33.0          0.0            1944.0          436.0   \n",
       "3     56.0    68.0         11.0          6.0            7081.0         2041.0   \n",
       "4    181.0    76.0         29.0         12.0           13850.0         1856.0   \n",
       "..     ...     ...          ...          ...               ...            ...   \n",
       "279   23.0     0.0          0.0         19.0            4187.0          199.0   \n",
       "280   80.0     0.0         14.0         11.0             481.0          245.0   \n",
       "281   25.0    59.0          0.0          0.0             408.0          258.0   \n",
       "282    0.0    16.0          0.0          0.0            4456.0          111.0   \n",
       "283    3.0    22.0          0.0          0.0             195.0           69.0   \n",
       "\n",
       "     ...  Rubinisphaera  Dictyoglomus  Pakpunavirus  Marinilactibacillus  \\\n",
       "0    ...            0.0           0.0           0.0                  0.0   \n",
       "1    ...            0.0           0.0           0.0                  0.0   \n",
       "2    ...            0.0           0.0           0.0                  0.0   \n",
       "3    ...            0.0           0.0           0.0                  0.0   \n",
       "4    ...            0.0           0.0           0.0                  0.0   \n",
       "..   ...            ...           ...           ...                  ...   \n",
       "279  ...            0.0           0.0           0.0                  0.0   \n",
       "280  ...            0.0           0.0           0.0                  0.0   \n",
       "281  ...            0.0           0.0           0.0                  0.0   \n",
       "282  ...            0.0           0.0           0.0                  0.0   \n",
       "283  ...            0.0           0.0           0.0                  0.0   \n",
       "\n",
       "     Paludibacter  Nonagvirus  Halovivax  Phifelvirus  Planktothrix  \\\n",
       "0             0.0         0.0        0.0          0.0           0.0   \n",
       "1             0.0         0.0        0.0          0.0           0.0   \n",
       "2             0.0         0.0        0.0          0.0           0.0   \n",
       "3             0.0         0.0        0.0          0.0           0.0   \n",
       "4             0.0         0.0        0.0          0.0           0.0   \n",
       "..            ...         ...        ...          ...           ...   \n",
       "279           0.0         0.0        0.0          0.0           0.0   \n",
       "280           0.0         0.0        0.0          0.0           0.0   \n",
       "281           0.0         0.0        0.0          0.0           0.0   \n",
       "282           0.0         0.0        0.0          0.0           0.0   \n",
       "283           0.0         0.0        0.0          0.0           0.0   \n",
       "\n",
       "     Denitrobacterium  \n",
       "0                 0.0  \n",
       "1                 0.0  \n",
       "2                 0.0  \n",
       "3                 0.0  \n",
       "4                 0.0  \n",
       "..                ...  \n",
       "279               0.0  \n",
       "280               0.0  \n",
       "281               0.0  \n",
       "282               0.0  \n",
       "283               0.0  \n",
       "\n",
       "[284 rows x 1058 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathogen</th>\n",
       "      <th>y</th>\n",
       "      <th>Saccharopolyspora</th>\n",
       "      <th>Lactococcus</th>\n",
       "      <th>Kocuria</th>\n",
       "      <th>Brevundimonas</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Propionibacterium</th>\n",
       "      <th>Oligella</th>\n",
       "      <th>Streptococcus</th>\n",
       "      <th>...</th>\n",
       "      <th>Brevibacterium</th>\n",
       "      <th>Alistipes</th>\n",
       "      <th>Buttiauxella</th>\n",
       "      <th>Bosea</th>\n",
       "      <th>Escherichia</th>\n",
       "      <th>Selenomonas</th>\n",
       "      <th>Lymphocryptovirus</th>\n",
       "      <th>Raoultella</th>\n",
       "      <th>Atopobium</th>\n",
       "      <th>Stenotrophomonas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>353.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>630.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>916.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>none</td>\n",
       "      <td>healthy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>367.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1416.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>737.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>2228.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2115.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Cryptococcus neoformans</td>\n",
       "      <td>septic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>614.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Streptococcus oralis</td>\n",
       "      <td>septic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26874.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>651.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1010.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>septic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>260924.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1104.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows × 230 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pathogen        y  Saccharopolyspora  Lactococcus  \\\n",
       "0                       none  healthy                0.0        133.0   \n",
       "1                       none  healthy                0.0         78.0   \n",
       "2                       none  healthy                0.0          0.0   \n",
       "3                       none  healthy                0.0          0.0   \n",
       "4                       none  healthy                0.0         45.0   \n",
       "..                       ...      ...                ...          ...   \n",
       "279         Escherichia coli   septic                0.0         84.0   \n",
       "280  Cryptococcus neoformans   septic                0.0         20.0   \n",
       "281     Streptococcus oralis   septic                0.0        136.0   \n",
       "282         Escherichia coli   septic                2.0          0.0   \n",
       "283         Escherichia coli   septic                0.0          0.0   \n",
       "\n",
       "     Kocuria  Brevundimonas  Acidaminococcus  Propionibacterium  Oligella  \\\n",
       "0       30.0           50.0              0.0                0.0       0.0   \n",
       "1       14.0           27.0              0.0                0.0       0.0   \n",
       "2        1.0           15.0              0.0                0.0       0.0   \n",
       "3        9.0           39.0              0.0                0.0       0.0   \n",
       "4      125.0           50.0              0.0                6.0       0.0   \n",
       "..       ...            ...              ...                ...       ...   \n",
       "279     57.0           48.0              0.0                0.0       0.0   \n",
       "280      0.0           40.0              0.0                0.0       0.0   \n",
       "281     59.0          127.0              0.0                0.0       0.0   \n",
       "282      0.0            2.0              0.0                0.0       0.0   \n",
       "283      0.0           10.0              0.0                0.0       0.0   \n",
       "\n",
       "     Streptococcus  ...  Brevibacterium  Alistipes  Buttiauxella  Bosea  \\\n",
       "0            227.0  ...             0.0        0.0           0.0  130.0   \n",
       "1            147.0  ...            18.0       15.0           0.0   60.0   \n",
       "2            115.0  ...             0.0        0.0           0.0   17.0   \n",
       "3            337.0  ...            34.0        0.0           0.0   56.0   \n",
       "4            367.0  ...             3.0       48.0           0.0  181.0   \n",
       "..             ...  ...             ...        ...           ...    ...   \n",
       "279          737.0  ...             0.0      190.0           0.0   23.0   \n",
       "280          614.0  ...             0.0        0.0           0.0   80.0   \n",
       "281        26874.0  ...             0.0        0.0           0.0   25.0   \n",
       "282           94.0  ...             0.0        0.0           0.0    0.0   \n",
       "283          356.0  ...             0.0        0.0           0.0    3.0   \n",
       "\n",
       "     Escherichia  Selenomonas  Lymphocryptovirus  Raoultella  Atopobium  \\\n",
       "0            0.0          0.0                0.0         0.0        0.0   \n",
       "1           11.0          0.0                0.0         0.0        0.0   \n",
       "2            0.0          0.0                0.0         0.0        0.0   \n",
       "3            0.0          0.0                0.0         0.0        0.0   \n",
       "4           87.0          0.0              187.0         0.0        0.0   \n",
       "..           ...          ...                ...         ...        ...   \n",
       "279       2228.0          0.0                0.0         0.0       14.0   \n",
       "280          2.0         27.0                0.0         0.0        0.0   \n",
       "281          0.0          0.0                0.0         0.0        0.0   \n",
       "282        293.0          0.0                0.0         0.0        0.0   \n",
       "283     260924.0          0.0                0.0         0.0        0.0   \n",
       "\n",
       "     Stenotrophomonas  \n",
       "0               353.0  \n",
       "1               987.0  \n",
       "2               630.0  \n",
       "3               916.0  \n",
       "4              1416.0  \n",
       "..                ...  \n",
       "279            2115.0  \n",
       "280            1131.0  \n",
       "281             651.0  \n",
       "282            1010.0  \n",
       "283            1104.0  \n",
       "\n",
       "[284 rows x 230 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "raw_df = pd.read_csv(datasets / 'karius_genus_raw.csv')\n",
    "pathogens_df = pd.read_csv(datasets / 'karius_genus_pathogens.csv')\n",
    "display(raw_df)\n",
    "display(pathogens_df)\n",
    "\n",
    "X = raw_df.iloc[:, 2:].copy()\n",
    "X_pathogens = pathogens_df.iloc[:, 2:].copy()\n",
    "y = raw_df.iloc[:, 1].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binary encode y\n",
    "y.loc[y == 'septic'] = 1\n",
    "y.loc[y == 'healthy'] = 0\n",
    "y = y.astype('int')\n",
    "\n",
    "# Relative abundance\n",
    "RA = X.apply(func=lambda x: x / x.sum(), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Test data\n",
    "X_idx, X_test_idx, y_train, y_test = train_test_split(X.index, y,\n",
    "                                                shuffle=True,\n",
    "                                                random_state=66,\n",
    "                                                test_size=0.3,\n",
    "                                                stratify=y)\n",
    "\n",
    "\n",
    "def get_datasets(df, train_idx, test_idx):\n",
    "    train = df.loc[train_idx, :]\n",
    "    test = df.loc[test_idx, :]\n",
    "    return train, test\n",
    "\n",
    "raw_train, raw_test = get_datasets(X, X_idx, X_test_idx)\n",
    "RA_train, RA_test = get_datasets(RA, X_idx, X_test_idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I print the number of examples for each split of data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_metadata(y, df):\n",
    "    pos = len(y[y == 1])\n",
    "    neg = len(y[y == 0])\n",
    "    row = pd.DataFrame({'Septic': [pos], 'Healthy': [neg]})\n",
    "    df = df.append(row)\n",
    "    return df\n",
    "\n",
    "\n",
    "metadata = pd.DataFrame(columns=['Septic', 'Healthy'])\n",
    "metadata = get_metadata(y_train, metadata)\n",
    "metadata = get_metadata(y_test, metadata)\n",
    "metadata.index = ['train', 'test']\n",
    "display(metadata)\n",
    "\n",
    "# Get negative to positive ratio\n",
    "ratio = sum(y == 0) / sum(y == 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimising model hyperparameters\n",
    "We use gridsearch to do hyperparameter optimisation (n_estimators, max_depth).\n",
    "n_estimators: Number of tree stumps\n",
    "max_depth: max depth of nodes in tree stumps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def optimise(X, y):\n",
    "    # Hyperparemeter Optimisation using grid search (F1)\n",
    "    model = XGBClassifier()\n",
    "    n_estimators = range(50, 300, 10)\n",
    "    max_depth = range(1, 5, 1)\n",
    "    gamma = [0.5, 1, 1.5, 2, 5]\n",
    "    subsample = [0.6, 0.7, 0.8, 0.9, 1.0]\n",
    "    colsample_bytree = np.linspace(0.1, 1, 10)\n",
    "    print(max_depth)\n",
    "    param_grid = dict(max_depth=max_depth, \n",
    "                      n_estimators=n_estimators, \n",
    "                      colsample_bytree=colsample_bytree,\n",
    "                      gamma = gamma,\n",
    "                     subsample = subsample,\n",
    "                     scale_pos_weight = [ratio])\n",
    "    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=66)\n",
    "    grid_search = GridSearchCV(model, param_grid, scoring=\"roc_auc\", n_jobs=-1, cv=kfold, verbose=1)\n",
    "    grid_result = grid_search.fit(X, y)\n",
    "    # summarize results\n",
    "    means = grid_result.cv_results_['mean_test_score']\n",
    "    stds = grid_result.cv_results_['std_test_score']\n",
    "    params = grid_result.cv_results_['params']\n",
    "    print(\"Best: %f using %s\" % (grid_result.best_score_, grid_result.best_params_))\n",
    "\n",
    "    # Print parameters and scores\n",
    "#     for mean, stdev, param in zip(means, stds, params):\n",
    "#         print(\"%f (%f) with: %r\" % (mean, stdev, param))\n",
    "\n",
    "    return(grid_result.best_params_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimise Model using Neat Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw_params = optimise(raw_train, y_train)\n",
    "raw_params = {'colsample_bytree': 0.1, 'gamma': 0.5, 'max_depth': 2, 'n_estimators': 140, 'scale_pos_weight': 1.4273504273504274, 'subsample': 0.6}\n",
    "\n",
    "# RA_params = optimise(RA_train, y_train)\n",
    "RA_params = {'colsample_bytree': 0.1, 'gamma': 1, 'max_depth': 1, 'n_estimators': 290, 'scale_pos_weight': 1.4273504273504274, 'subsample': 0.6}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_model = XGBClassifier(**raw_params)\n",
    "raw_model.fit(X=raw_train, y=y_train)\n",
    "\n",
    "RA_model = XGBClassifier(**RA_params)\n",
    "RA_model.fit(X=RA_train, y=y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluate Neat Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import precision_recall_fscore_support, roc_auc_score\n",
    "from imblearn.metrics import sensitivity_specificity_support\n",
    "\n",
    "def evaluate(model, X_Test, y_Test):\n",
    "    y_Pred = model.predict(X_Test)\n",
    "    y_Score = model.predict_proba(X_Test)[:, 1]\n",
    "\n",
    "    precision, recall, f1, _ = precision_recall_fscore_support(y_true=y_Test, y_pred=y_Pred, average='binary')\n",
    "    sensitivity, specificity, _ = sensitivity_specificity_support(y_true=y_Test, y_pred=y_Pred, average='binary')\n",
    "    auc = roc_auc_score(y_true=y_Test, y_score=y_Score)\n",
    "    metric_df = pd.DataFrame({'F1':[f1],\n",
    "                              'Sensitivity': [sensitivity],\n",
    "                              'Specificity': [specificity],\n",
    "                              'AUROC': [auc]})\n",
    "    return metric_df\n",
    "\n",
    "\n",
    "# Evaluate on test data\n",
    "raw_metric = evaluate(raw_model, raw_test, y_test)\n",
    "RA_metric = evaluate(RA_model, RA_test, y_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove Contaminants based on SHAP values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from scipy.stats import spearmanr\n",
    "import shap\n",
    "\n",
    "\n",
    "def decontam(X_train, y_train, params):\n",
    "    model = XGBClassifier(**params)\n",
    "    model.fit(X=X_train, y=y_train)\n",
    "\n",
    "    explainer = shap.TreeExplainer(model, feature_pertubation='interventional', model_output='probability', data=X_train)\n",
    "    shap_val = explainer.shap_values(X_train)\n",
    "\n",
    "    to_retain = np.array([True] * X_train.shape[1])\n",
    "    corrs = np.zeros(X_train.shape[1])\n",
    "\n",
    "    for i in range(X_train.shape[1]):\n",
    "        rho = spearmanr(X_train.iloc[:, i], shap_val[:, i])[0]\n",
    "        p = spearmanr(X_train.iloc[:, i], shap_val[:, i])[1]\n",
    "        if rho < 0 and p < 0.05:\n",
    "            to_retain[i] = False\n",
    "\n",
    "        if math.isnan(rho):\n",
    "            corrs[i] = 2\n",
    "        else:\n",
    "            corrs[i] = rho\n",
    "\n",
    "    to_retain = np.logical_and(corrs > 0, corrs != 2)\n",
    "    to_retain = X_train.columns[to_retain]\n",
    "    print(to_retain.shape, to_retain)\n",
    "    return to_retain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genera_new = raw_train.columns\n",
    "for _ in range(10):\n",
    "    genera_new = decontam(raw_train[genera_new], y_train, raw_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_retain = list(set(genera_new).intersection(set(X_pathogens.columns)))\n",
    "print(to_retain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Decontam + pathogens\n",
    "raw_SS = X_pathogens[to_retain]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get SHAP summary before removing Cellulomonas and Agrobacterium\n",
    "pre_SS_train, pre_SS_test = get_datasets(raw_SS, X_idx, X_test_idx)\n",
    "pre_model = XGBClassifier(**raw_params)\n",
    "pre_model.fit(X=pre_SS_train, y=y_train)\n",
    "\n",
    "pre_explainer = shap.TreeExplainer(pre_model, feature_pertubation='interventional', model_output='probability', data=pre_SS_train)\n",
    "shap_pre = pre_explainer.shap_values(pre_SS_train)\n",
    "\n",
    "shap.summary_plot(shap_pre, pre_SS_train, show=False, plot_size=(4, 5), color_bar_label='Unique k-mer Count', max_display=25)\n",
    "fig, ax = plt.gcf(), plt.gca()\n",
    "ax.set_xlabel('SHAP Value')\n",
    "plt.savefig(results / 'pre_shap.png', dpi=1200, format='png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_SS = raw_SS.drop(['Cellulomonas', 'Agrobacterium'], axis=1)\n",
    "print(f'Removed {len(genera_new) - len(to_retain)} genera')\n",
    "# print(f\"Genera from list not removed:\\n {set(meta) - set(to_remove)}\")\n",
    "print(f'New shape  = {raw_SS.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalise Datasets\n",
    "RA_SS = raw_SS.apply(func=lambda x: x / x.sum(), axis=1)\n",
    "\n",
    "# Get train test split\n",
    "raw_SS_train, raw_SS_test = get_datasets(raw_SS, X_idx, X_test_idx)\n",
    "RA_SS_train, RA_SS_test = get_datasets(RA_SS, X_idx, X_test_idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Number of Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Neat', X.shape)\n",
    "print('Pathogens', raw_SS.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimise SS Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathogens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw_SS_params = optimise(raw_SS_train, y_train)\n",
    "raw_SS_params = {'colsample_bytree': 0.1, 'gamma': 1, 'max_depth': 3, 'n_estimators': 270, 'scale_pos_weight': 1.4273504273504274, 'subsample': 1.0}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RA_SS_params = optimise(RA_SS_train, y_train)\n",
    "RA_SS_params = {'colsample_bytree': 0.2, 'gamma': 1, 'max_depth': 4, 'n_estimators': 70, 'scale_pos_weight': 1.4273504273504274, 'subsample': 0.6}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit optimised models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Fit optimised model on all training data\n",
    "\n",
    "# Decontam\n",
    "raw_SS_model = XGBClassifier(**raw_SS_params)\n",
    "raw_SS_model.fit(X=raw_SS_train, y=y_train)\n",
    "\n",
    "RA_SS_model = XGBClassifier(**RA_SS_params)\n",
    "RA_SS_model.fit(X=RA_SS_train, y=y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "raw_SS_metric = evaluate(raw_SS_model, raw_SS_test, y_test)\n",
    "RA_SS_metric = evaluate(RA_SS_model, RA_SS_test, y_test)\n",
    "\n",
    "raw_SS_metric = evaluate(raw_SS_model, raw_SS_test, y_test)\n",
    "RA_SS_metric = evaluate(RA_SS_model, RA_SS_test, y_test)\n",
    "\n",
    "metric_df = pd.concat([raw_metric,\n",
    "                       RA_metric,\n",
    "                       raw_SS_metric,\n",
    "                       RA_SS_metric], axis=0)\n",
    "metric_df.index = ['Raw', 'RA', 'Raw SS', 'RA SS']\n",
    "display(metric_df.round(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Confidence Intervals (non-parametric boostrap estimates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bootstrap with 1001 iterations, 95% CI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_percentiles(x, alpha=0.05):\n",
    "    low = np.percentile(x, alpha / 2 * 100)\n",
    "    high = np.percentile(x, (1 - alpha / 2) * 100)\n",
    "    \n",
    "    return low, high\n",
    "\n",
    "\n",
    "np.random.seed(66)\n",
    "from sklearn.utils import resample\n",
    "\n",
    "\n",
    "def get_confint(model, X_test, y_test, n_iter=1001):\n",
    "    boot_df = pd.DataFrame({'F1': [0], 'Sensitivity': [0], 'Specificity': [0], 'AUROC' : [0]})\n",
    "    \n",
    "    for _ in range(n_iter):\n",
    "        boot_X, boot_y = resample(X_test, y_test, n_samples=len(y_test), replace=True, stratify=y_test)\n",
    "        y_pred = model.predict(boot_X)\n",
    "        y_score = model.predict_proba(boot_X)[:, 1]\n",
    "\n",
    "        sensitivity, specificity, _ = sensitivity_specificity_support(y_true=boot_y, y_pred=y_pred, average='binary')\n",
    "        precision, recall, f1, _ = precision_recall_fscore_support(y_true=boot_y, y_pred=y_pred, average='binary')\n",
    "        auc = roc_auc_score(y_true=boot_y, y_score=y_score)\n",
    "        temp_df = pd.DataFrame({'F1': [f1], 'Sensitivity': [sensitivity], \n",
    "                                'Specificity': [specificity], 'AUROC' : [auc]})\n",
    "        \n",
    "        boot_df = pd.concat([boot_df, temp_df], axis=0)\n",
    "    \n",
    "    boot_df = boot_df.iloc[1:, :]\n",
    "    \n",
    "    confints = [get_percentiles(boot_df[col]) for col in boot_df.columns]\n",
    "    display(pd.DataFrame(confints, \n",
    "                         columns=['2.5%', '97.5%'], \n",
    "                         index=boot_df.columns).transpose().round(3))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Raw:', end='')\n",
    "get_confint(raw_model, raw_test, y_test)\n",
    "print('RA:', end='')\n",
    "get_confint(RA_model, RA_test, y_test)\n",
    "\n",
    "print('Raw SS:', end='')\n",
    "get_confint(raw_SS_model, raw_SS_test, y_test)\n",
    "print('RA SS:', end='')\n",
    "get_confint(RA_SS_model, RA_SS_test, y_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interpreting model using SHAP values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Feature importance\n",
    "This is a plot of mean absolute SHAP values per feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot of SHAP values per Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "explainer_SS = shap.TreeExplainer(raw_SS_model, feature_pertubation='interventional', model_output='probability', data=raw_SS_train)\n",
    "shap_SS = explainer_SS.shap_values(raw_SS_test)\n",
    "\n",
    "explainer_raw = shap.TreeExplainer(raw_model, feature_pertubation='interventional', model_output='probability', data=raw_train)\n",
    "shap_raw = explainer_raw.shap_values(raw_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_SS, raw_SS_test, show=False, plot_size=(4, 5), color_bar_label='Unique k-mer Count', max_display=35)\n",
    "fig, ax = plt.gcf(), plt.gca()\n",
    "ax.set_xlabel('SHAP Value')\n",
    "plt.savefig(results / 'SS_shap.png', dpi=1200, format='png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shap.summary_plot(shap_raw, raw_test, show=False, plot_size=(4, 5), color_bar_label='Unique k-mer Count', max_display=23)\n",
    "fig, ax = plt.gcf(), plt.gca()\n",
    "ax.set_xlabel('SHAP Value')\n",
    "plt.savefig(results / 'raw_shap.png', dpi=1200, format='png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Features are ranked by importance from top to botttom\n",
    "* feature values are the kmer counts for each genus\n",
    "* SHAP values are the average marginal contributions to probability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Force plot for healthy patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "j = 72\n",
    "print(f'Actual Classification {y_test[j]}')\n",
    "print(raw_SS_test.index[j])\n",
    "shap.force_plot(explainer_SS.expected_value, \n",
    "                shap_SS[j,:], \n",
    "                raw_SS_test.iloc[j,:],\n",
    "                show=False,\n",
    "                matplotlib=True)\n",
    "plt.savefig(results / 'SS_force_plot.png', dpi=1200, format='png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How much does Escherichia drive predictions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "escherichia_idx = raw_SS_test.columns.get_loc('Escherichia')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_score = raw_SS_model.predict_proba(raw_SS_test)[:, 1]\n",
    "old_auc = roc_auc_score(y_true=y_test, y_score=y_score)\n",
    "new_auc = roc_auc_score(y_true=y_test, y_score=y_score - shap_SS[:, escherichia_idx])\n",
    "print(f\"Before Removing Escherichia = {old_auc}\\nAfter Removing Escherichia = {new_auc}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mashin",
   "language": "python",
   "name": "mashin"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
